### DISCUSSION TITLE
How do I specify imagePullSecret for the task/workflow pods?

### DISCUSSION THREAD

**Discussion by kumare3** [flyteorg/flyte#1107] [Q&A] [answered]
What is the correct way to specify imagePullSecret for the task/workflow pods? I'm trying to use a private Docker registry and currently get the "Container with unready status" error

**Comment by kumare3** [comment, marked as answer]
There are a few ways to do this,
- `Manually`: Update the default service account (or if using a specific serviceaccount then update that account) to add the imagePullSecrets. As explained [here](https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/)
- `Automatically`: For every new project added to Flyte, Flyte runs a specific component called the `ClusterResourceManager`, which is responsible to keep the cluster in sync with these various requirements. This is an optional component. It essentially applies some K8s resource with a few templatable parameters to all namespaces that flyte will use. 
              - For Helm an example can  be seen [here](https://github.com/flyteorg/flyte/blob/master/helm/values-eks.yaml#L380-L389). The example shows how to patch the default service account. 
              - For kustomize similarly - Add the template [here](https://github.com/flyteorg/flyte/tree/master/kustomize/overlays/sandbox/flyte/config/clusterresource-templates) and refer to it [here](https://github.com/flyteorg/flyte/blob/master/kustomize/overlays/sandbox/flyte/kustomization.yaml#L28).
- Lastly: You can use the [K8sPod plugin](https://docs.flyte.org/projects/cookbook/en/latest/auto/integrations/kubernetes/pod/index.html) (this is a very high quality almost core plugin in Flyte) to customize your pods in anyway. This executes the entire pod spec.

```
Note: If using cluster resource manager, then you can patch the default serviceaccount. But if you are using workload identity or IAM for service accounts, Flyte cannot create these service accounts and you should manage them outside of Flyte (like in a CI system, terraform or through console). This is where you should add the imagepullsecret and then just pass the serviceaccount to a flyte launch plan or when starting an execution.
```

**Reply by maximsmol** [reply]
What worked for me was patching the `default` service account ❗ but under the `PROJECTNAME-development` 
namespace. The `imagePullSecrets` entry must, of course, also refer to a secret under the same namespace for this to work.

**Note:** I had to create the project namespace manually (just yet another Kustomize resource file) since otherwise k8s will refuse to apply the config. Flyte has no issues with the existing namespace when registering the project later.
